Source/WebGPU/WGSL/GlobalVariableRewriter.cpp

```diff
@@ -0,0 +1,281 @@
+    m_callGraph.ast().globalVars().clear();
+}
+
+void RewriteGlobalVariables::visit(AST::FunctionDecl& functionDecl)
+{
+    // FIXME: visit callee once we have any
+
+    for (auto& parameter : functionDecl.parameters())
+        def(parameter.name());
+
+    // FIXME: detect when we shadow a global that a callee needs
+    AST::Visitor::visit(functionDecl.body());
+}
+
+void RewriteGlobalVariables::visit(AST::VariableDecl& variableDecl)
+{
+    def(variableDecl.name());
+}
+
+template <typename TargetType, typename CurrentType, typename... Arguments>
+void replace(CurrentType& dst, Arguments&&... arguments)
+{
+    static_assert(sizeof(TargetType) <= sizeof(CurrentType));
+    new (&dst) TargetType(dst.span(), std::forward<Arguments>(arguments)...);
+}
+
+void RewriteGlobalVariables::visit(AST::IdentifierExpression& identifier)
+{
+    auto name = identifier.identifier();
+    if (Global* global = read(name)) {
+        if (auto resource = global->m_resource) {
+            replace<AST::StructureAccess>(identifier,
+                makeUniqueRef<AST::IdentifierExpression>(identifier.span(), argumentBufferParameterName(resource->m_group)),
+                WTFMove(name));
+        }
+    }
+}
+
+void RewriteGlobalVariables::collectGlobals()
+{
+    auto& globalVars = m_callGraph.ast().globalVars();
+    for (auto& globalVar : globalVars) {
+        std::optional<unsigned> group;
+        std::optional<unsigned> binding;
+        for (auto& attribute : globalVar.attributes()) {
+            if (attribute->kind() == AST::Node::Kind::GroupAttribute) {
+                group = { downcast<AST::GroupAttribute>(attribute.get()).group() };
+                continue;
+            }
+            if (attribute->kind() == AST::Node::Kind::BindingAttribute) {
+                binding = { downcast<AST::BindingAttribute>(attribute.get()).binding() };
+                continue;
+            }
+        }
+        std::optional<Global::Resource> resource;
+        if (group.has_value()) {
+            RELEASE_ASSERT(binding.has_value());
+            resource = { *group, *binding };
+        }
+
+        auto result = m_globals.add(globalVar.name(), Global {
+            resource,
+            &globalVar
+        });
+        ASSERT(result, result.isNewEntry);
+
+        if (resource.has_value()) {
+            Global& global = result.iterator->value;
+            auto result = m_groupBindingMap.add(resource->m_group, IndexMap<Global*>());
+            result.iterator->value.add(resource->m_binding, &global);
+        }
+    }
+}
+
+void RewriteGlobalVariables::visitEntryPoint(AST::FunctionDecl& functionDecl)
+{
+    m_reads.clear();
+    m_defs.clear();
+
+    visit(functionDecl);
+    if (m_reads.isEmpty())
+        return;
+
+    auto groupsToGenerate = requiredGroups();
+    // FIXME: not all globals are parameters
+    insertParameters(functionDecl, groupsToGenerate);
+}
+
+
+auto RewriteGlobalVariables::requiredGroups() -> IndexSet
```

Comment:
This style isn't consistent with the rest of the code. I tried it at one point and @litherum questioned it. I'm happy to use either syntax but we should be consistent.

Reply:
The purpose of the `->` is for namespacing.

`int Foo::myThingy()` is better than `auto myThingy -> int`

`Foo::Bar Foo::myThingy()` is worse than `auto Foo::myThingy() -> Bar`

Reply:
https://github.com/WebKit/WebKit/pull/9200